src/main/java/ch/njol/skript/sections/SecRecipe.java

```diff
@@ -0,0 +1,151 @@
+		result = (Expression<ItemType>) exprs[0];
+		key = (Expression<String>) exprs[1];
+		loadCode(sectionNode);
+		if (first == null) {
+			Skript.error("A recipe section must contain 3 recipe lines");
+			return false;
+		}
+		TriggerItem current = first;
+		for (int i = 0; i < 3; i++) {
+			if (current == null || current.getClass() != EffRecipeLine.class) {
+				Skript.error("A recipe section may only contain 3 recipe lines - 1");
+				return false;
+			}
+			current = current.getNext();
+		}
+		if (Iterables.size(sectionNode) != 3) {
+			Skript.error("A recipe section may only contain 3 recipe lines- 2");
+			return false;
+		}
+		return true;
+	}
+
+	private ItemType[] getAllIngredients(Event event) {
+		EffRecipeLine firstLine = (EffRecipeLine) first;
+		EffRecipeLine secondLine = (EffRecipeLine) firstLine.getNext();
+		EffRecipeLine thirdLine = (EffRecipeLine) secondLine.getNext();
+		return Stream.of(firstLine.getIngredients(event), secondLine.getIngredients(event), thirdLine.getIngredients(event))
+			.flatMap(Arrays::stream)
+			.toArray(ItemType[]::new);
+	}
+
+	@Override
+	public String toString(Event event, boolean debug) {
+		return "create crafting recipe for " + result.toString(event, debug) + " with key " + key.toString(event, debug);
+	}
+
+	@Override
+	protected TriggerItem walk(Event event) {
+		execute(event);
+		return walk(event, false);
+	}
+
+	protected void execute(Event event) {
+		String key = this.key.getSingle(event);
+		if (key == null)
+			return;
+		ItemType result = this.result.getSingle(event);
+		if (result == null)
+			return;
+		ItemType[] ingredients = getAllIngredients(event);
+		if (ingredients.length != 9)
+			return;
+		ShapedRecipe recipe = new ShapedRecipe(new NamespacedKey(Skript.getInstance(), key), result.getRandom());
+		recipe.shape("abc", "def", "ghi");
+		for (char c = 'a'; c < 'j'; c++) {
+			recipe.setIngredient(c, ingredients[c - 'a'].getMaterial());
+		}
+		try {
+			Bukkit.getServer().addRecipe(recipe);
+		} catch (IllegalStateException ignored) {
+			// Bukkit throws a IllegalStateException if a duplicate recipe is registered
+		}
+	}
+
+
+	public static class EffRecipeLine extends Effect {
+
+		private Expression<ItemType> ingredients;
+
+		@SuppressWarnings("unchecked")
+		@Override
+		public boolean init(Expression<?>[] exprs, int matchedPattern, Kleenean isDelayed, SkriptParser.ParseResult parseResult) {
+			ingredients = (Expression<ItemType>) exprs[0];
+			if (!(ingredients instanceof ExpressionList)) {
+				Skript.error("A recipe line may only contain a list of items");
+				return false;
+			}
+			if (((ExpressionList<ItemType>) ingredients).getExpressions().length != 3) {
+				Skript.error("An recipe line may only contain three items");
+			}
+			List<TriggerSection> currentSections = getParser().getCurrentSections();
+			if (currentSections.isEmpty())
+				return false;
+			return currentSections.get(currentSections.size() - 1).getClass() == SecRecipe.class;
```

Comment:
check the section first. we don't want to print errors if we're in the wrong section